### PR TITLE
[WIP] added Slidable widget

### PR DESCRIPTION
**Pull request by nt4f04uNd**
adds Slidable widget - it allows to define some "slide" gesture behavior for some widget, so it will slide from one position to another

i chose "slide" over "swipe" here because slide more precisely describes this gesture. swipe is a single fast gesture, whilst slide is that you put finger down and also can drag the widget in both directions

### use cases

basically, it generalizes a lot of places where such behavior is used. it might be used to simplify a lot of widgets framework already has:

* Dismissible
* BottomSheets
* Drawer
* iOS swipe to back

it can help implementing these issues https://github.com/flutter/flutter/issues/33798, https://github.com/flutter/flutter/issues/42560 and other similar

and lastly - it just would allow developers to write their own widget that require such behavior in a really fast and easy way

### notes

to allow easily try it out, i made a few examples in https://github.com/nt4f04unds-archive/slidable_example
it's now blocked by https://github.com/flutter/flutter/pull/76017, so checkout into https://github.com/nt4f04uNd/flutter/tree/slidable_example to build it

i just made a really basic implementation - it doesn't have any tests, bearly documented. i would like first for someone to take a look at it and tell me what he thinks about it, because i don't want to spend time on a feature that might be eventually rejected

in future i plan:
* add snapping points
* support non-fraction drag and offsets
* split AnimationStrategy into other PR
* some other stuff

example app videos

https://user-images.githubusercontent.com/39104740/108003700-952f0980-7004-11eb-9108-8f171b69514c.mp4

https://user-images.githubusercontent.com/39104740/108003708-97916380-7004-11eb-99e1-9ec7fe0d1e46.mp4

https://user-images.githubusercontent.com/39104740/108003713-995b2700-7004-11eb-96d4-537786e7fb1b.mp4

### CODE REVIEW

**Comment by nt4f04uNd** [comment]
@Hixie (or someone else) if you could take a short glance at it and tell me what you think about this concept, it would be extra cool

**Comment by Hixie** [comment]
Would it make sense for this to be a package? (Like package:animation, for instance?)

**Comment by nt4f04uNd** [comment]
maybe, but its main idea is not that it animates a widget, it implements mentioned slide gesture combined with animating it and that's why i put it into widgets. do you think it better suits package:animation?

and what do you think in general, do you like/dislike it and abilities it will bring to framework?

**Comment by Hixie** [comment]
To be honest I'm not sure I 100% understand what problem this PR is solving.

I didn't mean necessarily that it should be in package:animation, I just mean that in general we add new features to packages rather than to the core framework when we can (e.g. as we did with package:animation). If it's possible to create this as a package then that is probably the right solution. On the other hand if it's something that is only possible by integrating it with the framework, then that's a different matter.

**Comment by nt4f04uNd** [comment]
re: ***I didn't mean necessarily that it should be in package:animation***

so, i guess you meant package:animations, not animation which is a part of the core

re: ***To be honest I'm not sure I 100% understand what problem this PR is solving.***

refer to OP

> basically, it generalizes a lot of places where such behavior is used

i think these are three primary reasons to have it in the core and not in some separate package

1. this is something really really basic
2. it could simplify a lot of widgets in the framewrok itself as they are using such gestrure
3. it could help implementing new features

by the way i'm not 100% sure if it should be called Slidable, but Draggable already exists and Swipable seems to be misleading

**Comment by Hixie** [comment]
Let me rephrase my question. What problem does this PR solve?

**Comment by nt4f04uNd** [comment]
repetitivity of the code that implements such gesture

**Comment by Hixie** [comment]
Ok. If the problem is how to reduce repetitive work then making a package would be the perfect solution. That way the package can evolve and change and its evolution is not tied to the framework version.

**Comment by nt4f04uNd** [comment]
do i understand correctly that you think framework couldn't benefit from this widget per listed bullets?

> * it could simplify a lot of widgets in the framewrok itself as they are using such gestrure
> * it could help implementing new features

**Comment by Hixie** [comment]
That's certainly possible, but we'd want to first prove out the feature and get experience with it in a package.

We try to be very conservative with what we put in the core framework, because there's a high cost to having anything there. We have to commit to supporting it for years to come, we have to document it, test it, create samples, we have to consider everyone's varied desires which they may have as they use the feature, we have to fix bugs. If there's design problems, we may not find out for a long time but then once we do we then have to figure out how to fix them without breaking people, or we have to migrate all our existing widgets to the new architecture, etc.

Basically, code is expensive. So before we take it, we need to prove it. By creating a package, we can see if people use the feature, how they like it, whether it would be useful for the framework, etc, without having to take on any of the costs.

**Comment by nt4f04uNd** [comment]
you should have explained it this way prior. so, i should PR this as separate package, as part of `animations`, or make my own?

**Comment by Hixie** [comment]
Sorry for not being clear. I'll update the relevant section of the style guide to be clearer (https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#deciding-where-to-put-code).

I would probably start with an independent package.
